File: backend/app/utils/slug.py
"""URL slug utilities."""
import re
from unicodedata import normalize


def slugify(text: str) -> str:
    """
    Convert text to URL-friendly slug.

    Examples:
        "Så mycket kostar bensin idag" -> "sa-mycket-kostar-bensin-idag"
        "Här är Sveriges största städer!" -> "har-ar-sveriges-storsta-stader"
    """
    if not text:
        return ""

    # Normalize unicode characters
    text = normalize('NFKD', text).encode('ascii', 'ignore').decode('ascii')

    # Convert to lowercase
    text = text.lower()

    # Replace Swedish characters before removing
    text = text.replace('å', 'a').replace('ä', 'a').replace('ö', 'o')
    text = text.replace('é', 'e').replace('è', 'e').replace('ü', 'u')

    # Remove non-alphanumeric characters (except spaces and hyphens)
    text = re.sub(r'[^a-z0-9\s-]', '', text)

    # Replace spaces and multiple hyphens with single hyphen
    text = re.sub(r'[\s-]+', '-', text)

    # Remove leading/trailing hyphens
    text = text.strip('-')

    # Limit length
    if len(text) > 100:
        text = text[:100].rsplit('-', 1)[0]

    return text

File: frontend/src/utils/slug.ts
/**
 * Convert text to URL-friendly slug.
 */
export function slugify(text: string): string {
  if (!text) return ''

  let slug = text
    .toLowerCase()
    .normalize('NFD')
    .replace(/[\u0300-\u036f]/g, '') // Remove diacritics
    .replace(/å/g, 'a')
    .replace(/ä/g, 'a')
    .replace(/ö/g, 'o')
    .replace(/[^a-z0-9\s-]/g, '') // Remove non-alphanumeric except spaces and hyphens
    .replace(/[\s-]+/g, '-') // Replace spaces and multiple hyphens with single hyphen
    .replace(/^-+|-+$/g, '') // Remove leading/trailing hyphens

  // Only truncate and remove incomplete word if text is too long
  if (slug.length > 100) {
    slug = slug.substring(0, 100)
    const lastDash = slug.lastIndexOf('-')
    if (lastDash > 0) {
      slug = slug.substring(0, lastDash)
    }
  }

  return slug
}

/**
 * Generate article URL from date and title.
 */
export function generateArticleUrl(date: string, title: string): string {
  const dateObj = new Date(date)
  const dateStr = dateObj.toISOString().split('T')[0] // YYYY-MM-DD
  const slug = slugify(title)

  return `/articles/${dateStr}/${slug}`
}
